File: src/shared/sources/winnow/culling.ts
/**
 * Winnow's CULLING, as Atelier reads it — never writes it.
 *
 * Culling is Winnow's job (`docs/develop-tool.md` §8), and the maintainer's
 * answer of 2026-09-23 (`docs/lightroom-gaps.md` §8, item 33) is that Develop
 * SHOWS it and FILTERS on it, and never sets it. So what is here is a reader of
 * the row Winnow already sends (`GRID_SELECT` joins `ratings` into every row
 * of `/api/assets`) and a filter over what it read.
 *
 * - **The verdict** is Winnow's own vocabulary, checked against its schema
 *   (migrations 0001 and 0016): `pick`, `reject`, `skip` (swiped past in
 *   Sift) and `unrated`. Anything else reads as `unrated` rather than as a
 *   verdict nobody gave.
 * - **Stars** are 0 to 5; a value off that scale is clamped, not trusted.
 * - **A colour label** is a free string on Winnow's side — no screen of its
 *   own sets it today. It is shown only when it is one of the five names
 *   Lightroom and Capture One share; any other text is kept for the tooltip.
 *
 * A picture that did not come from an instance, or whose instance has not
 * answered, has NO culling (`undefined`) — which is not the same as unrated,
 * and the filter says so rather than hiding it as if it had been rejected.
 *
 * Pure and DOM-free.
 */

export type Verdict = 'pick' | 'reject' | 'skip' | 'unrated';

export interface Culling {
  verdict: Verdict;
  /** 0 to 5; 0 is no stars. */
  star: number;
  /** The label as Winnow holds it, or null for none. */
  color: string | null;
}

const VERDICTS: readonly Verdict[] = ['pick', 'reject', 'skip', 'unrated'];

/** The culling a Winnow row carries, or null when the row says nothing of it (an older instance). */
export function cullingFromRow(row: { verdict?: unknown; star?: unknown; color_label?: unknown }): Culling | null {
  if (row.verdict === undefined && row.star === undefined && row.color_label === undefined) return null;
  const verdict = VERDICTS.includes(row.verdict as Verdict) ? (row.verdict as Verdict) : 'unrated';
  const star = typeof row.star === 'number' && Number.isFinite(row.star) ? Math.max(0, Math.min(5, Math.round(row.star))) : 0;
  const color = typeof row.color_label === 'string' && row.color_label.trim() ? row.color_label.trim() : null;
  return { verdict, star, color };
}

/** True when Winnow has said anything about this picture beyond "not looked at". */
export function isCulled(c: Culling | null | undefined): boolean {
  return Boolean(c && (c.verdict !== 'unrated' || c.star > 0 || c.color));
}

/** The five labels Lightroom and Capture One share, by the name Winnow would store. */
export const LABEL_COLOURS = ['red', 'yellow', 'green', 'blue', 'purple'] as const;
export type LabelColour = (typeof LABEL_COLOURS)[number];

/** A stored label as one of the five, or null for none or a free text. */
export function labelColour(color: string | null | undefined): LabelColour | null {
  const c = color?.trim().toLowerCase();
  return c && (LABEL_COLOURS as readonly string[]).includes(c) ? (c as LabelColour) : null;
}

/** `Pick · ★★★ · red` — what Winnow said, in a line; empty for nothing. */
export function describeCulling(c: Culling | null | undefined): string {
  if (!c) return '';
  const parts: string[] = [];
  if (c.verdict === 'pick') parts.push('Pick');
  else if (c.verdict === 'reject') parts.push('Rejected');
  else if (c.verdict === 'skip') parts.push('Skipped');
  if (c.star > 0) parts.push('★'.repeat(c.star));
  if (c.color) parts.push(c.color);
  return parts.join(' · ');
}

// --- the filter ----------------------------------------------------------------

/**
 * What the strip shows, on Winnow's word. `all` is no filter; `picks` is
 * Winnow's picks alone; `unrejected` leaves the rejects out and keeps
 * everything else, a picture Winnow never saw included; `stars` keeps a
 * picture with at least `min` stars.
 */
export type CullFilter =
  | { kind: 'all' }
  | { kind: 'picks' }
  | { kind: 'unrejected' }
  | { kind: 'stars'; min: number };

export const NO_CULL_FILTER: CullFilter = { kind: 'all' };

/** Every filter the menu offers, in its order. */
export const CULL_FILTERS: readonly CullFilter[] = [
  { kind: 'all' },
  { kind: 'picks' },
  { kind: 'unrejected' },
  ...[1, 2, 3, 4, 5].map((min) => ({ kind: 'stars' as const, min })),
];

export function cullFilterKey(f: CullFilter): string {
  return f.kind === 'stars' ? `stars:${f.min}` : f.kind;
}

/** A stored key read back; anything unknown is no filter. */
export function readCullFilter(key: unknown): CullFilter {
  return CULL_FILTERS.find((f) => cullFilterKey(f) === key) ?? NO_CULL_FILTER;
}

export function cullFilterLabel(f: CullFilter): string {
  switch (f.kind) {
    case 'all':
      return 'All';
    case 'picks':
      return 'Picks';
    case 'unrejected':
      return 'Not rejected';
    case 'stars':
      return `${'★'.repeat(f.min)}${f.min < 5 ? ' and up' : ''}`;
  }
}

/**
 * Whether a picture passes. A picture Winnow has said NOTHING about
 * (`undefined`: from this computer, or its instance not answering) passes
 * only the filters that do not ask Winnow for a yes — `all` and
 * `unrejected` — because nothing says it was rejected, and nothing says it
 * was picked either.
 */
export function passesCull(c: Culling | null | undefined, f: CullFilter): boolean {
  switch (f.kind) {
    case 'all':
      return true;
    case 'unrejected':
      return c?.verdict !== 'reject';
    case 'picks':
      return c?.verdict === 'pick';
    case 'stars':
      return (c?.star ?? 0) >= f.min;
  }
}

/** How a roll's pictures stand against Winnow, for the status line. */
export interface CullCounts {
  /** Pictures Winnow answered for. */
  known: number;
  picks: number;
  rejects: number;
  starred: number;
}

export function countCulling(cullings: Iterable<Culling | null | undefined>): CullCounts {
  const out: CullCounts = { known: 0, picks: 0, rejects: 0, starred: 0 };
  for (const c of cullings) {
    if (!c) continue;
    out.known += 1;
    if (c.verdict === 'pick') out.picks += 1;
    if (c.verdict === 'reject') out.rejects += 1;
    if (c.star > 0) out.starred += 1;
  }
  return out;
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Sources/Winnow/Culling.swift
// Winnow's CULLING, as Atelier reads it — never writes it. Port of
// `src/shared/sources/winnow/culling.ts`.
//
// Culling is Winnow's job (`docs/develop-tool.md` §8), and the maintainer's
// answer of 2026-09-23 (`docs/lightroom-gaps.md` §8, item 33) is that Develop
// SHOWS it and FILTERS on it, and never sets it. So this is a reader of the
// row Winnow already sends (`GRID_SELECT` joins `ratings` into every row of
// `/api/assets`) and a filter over what it read.
//
// - The verdict is Winnow's vocabulary (its migrations 0001 and 0016):
//   `pick`, `reject`, `skip` and `unrated`; anything else reads as unrated,
//   never as a verdict nobody gave.
// - Stars are 0 to 5; a value off that scale is clamped, not trusted.
// - A colour label is a free string on Winnow's side; it is SHOWN only when
//   it is one of the five names Lightroom and Capture One share.
//
// A picture that did not come from an instance, or whose instance has not
// answered, has NO culling (nil) — which is not unrated, and the filter says
// so rather than hiding it as if it had been rejected.

import Foundation

public enum Verdict: String, Equatable, Sendable, CaseIterable {
    case pick, reject, skip, unrated
}

public struct Culling: Equatable, Sendable {
    public var verdict: Verdict
    /// 0 to 5; 0 is no stars.
    public var star: Int
    /// The label as Winnow holds it, or nil for none.
    public var color: String?

    public init(verdict: Verdict, star: Int, color: String?) {
        self.verdict = verdict; self.star = star; self.color = color
    }
}

/// The culling a row carries — `{ verdict, star, color_label }` — or nil when
/// the row says nothing of it (an older instance). A key that is ABSENT and a
/// key that is `null` differ: a row with `verdict: null` has been answered for.
public func cullingFromRow(_ raw: JSONValue?) -> Culling? {
    let o = raw?.objectValue ?? [:]
    return cullingFrom(verdict: o["verdict"], star: o["star"], colorLabel: o["color_label"])
}

/// The same reading of a row already read (`WinnowAssetRow` keeps the three
/// wire values as they came).
public func cullingFromRow(_ row: WinnowAssetRow) -> Culling? {
    cullingFrom(verdict: row.verdict, star: row.star, colorLabel: row.colorLabel)
}

private func cullingFrom(verdict: JSONValue?, star: JSONValue?, colorLabel: JSONValue?) -> Culling? {
    if verdict == nil && star == nil && colorLabel == nil { return nil }
    let v = verdict?.stringValue.flatMap(Verdict.init(rawValue:)) ?? .unrated
    var stars = 0
    if let n = star?.finiteNumber {
        // `Math.round`: half UP.
        let rounded = (n + 0.5).rounded(.down)
        stars = Int(Swift.max(0, Swift.min(5, rounded)))
    }
    let label = colorLabel?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines)
    return Culling(verdict: v, star: stars, color: (label?.isEmpty ?? true) ? nil : label)
}

/// True when Winnow has said anything about this picture beyond "not looked at".
public func isCulled(_ c: Culling?) -> Bool {
    guard let c else { return false }
    return c.verdict != .unrated || c.star > 0 || c.color != nil
}

/// The five labels Lightroom and Capture One share, by the name Winnow would store.
public enum LabelColour: String, Equatable, Sendable, CaseIterable {
    case red, yellow, green, blue, purple
}

/// A stored label as one of the five, or nil for none or a free text.
public func labelColour(_ color: String?) -> LabelColour? {
    guard let c = color?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(), !c.isEmpty else { return nil }
    return LabelColour(rawValue: c)
}

/// `Pick · ★★★ · red` — what Winnow said, in a line; empty for nothing.
public func describeCulling(_ c: Culling?) -> String {
    guard let c else { return "" }
    var parts: [String] = []
    switch c.verdict {
    case .pick: parts.append("Pick")
    case .reject: parts.append("Rejected")
    case .skip: parts.append("Skipped")
    case .unrated: break
    }
    if c.star > 0 { parts.append(String(repeating: "★", count: c.star)) }
    if let color = c.color { parts.append(color) }
    return parts.joined(separator: " · ")
}

// MARK: - the filter

/// What the strip shows, on Winnow's word. `all` is no filter; `picks` is
/// Winnow's picks alone; `unrejected` leaves the rejects out and keeps
/// everything else, a picture Winnow never saw included; `stars` keeps a
/// picture with at least `min` stars.
public enum CullFilter: Equatable, Sendable {
    case all
    case picks
    case unrejected
    case stars(min: Int)
}

/// The web's `NO_CULL_FILTER`.
public let noCullFilter: CullFilter = .all

/// Every filter the menu offers, in its order — the web's `CULL_FILTERS`.
public let cullFilters: [CullFilter] = [.all, .picks, .unrejected] + (1...5).map { CullFilter.stars(min: $0) }

public func cullFilterKey(_ f: CullFilter) -> String {
    switch f {
    case .all: return "all"
    case .picks: return "picks"
    case .unrejected: return "unrejected"
    case .stars(let min): return "stars:\(min)"
    }
}

/// A stored key read back; anything unknown is no filter.
public func readCullFilter(_ key: JSONValue?) -> CullFilter {
    guard let text = key?.stringValue else { return noCullFilter }
    return cullFilters.first { cullFilterKey($0) == text } ?? noCullFilter
}

public func cullFilterLabel(_ f: CullFilter) -> String {
    switch f {
    case .all: return "All"
    case .picks: return "Picks"
    case .unrejected: return "Not rejected"
    case .stars(let min): return String(repeating: "★", count: Swift.max(0, min)) + (min < 5 ? " and up" : "")
    }
}

/// Whether a picture passes. A picture Winnow has said NOTHING about (nil)
/// passes only the filters that do not ask Winnow for a yes — `all` and
/// `unrejected` — because nothing says it was rejected, nor picked.
public func passesCull(_ c: Culling?, _ f: CullFilter) -> Bool {
    switch f {
    case .all: return true
    case .unrejected: return c?.verdict != .reject
    case .picks: return c?.verdict == .pick
    case .stars(let min): return (c?.star ?? 0) >= min
    }
}

/// How a roll's pictures stand against Winnow, for the status line.
public struct CullCounts: Equatable, Sendable {
    /// Pictures Winnow answered for.
    public var known: Int
    public var picks: Int
    public var rejects: Int
    public var starred: Int

    public init(known: Int = 0, picks: Int = 0, rejects: Int = 0, starred: Int = 0) {
        self.known = known; self.picks = picks; self.rejects = rejects; self.starred = starred
    }
}

public func countCulling<S: Sequence>(_ cullings: S) -> CullCounts where S.Element == Culling? {
    var out = CullCounts()
    for case let c? in cullings {
        out.known += 1
        if c.verdict == .pick { out.picks += 1 }
        if c.verdict == .reject { out.rejects += 1 }
        if c.star > 0 { out.starred += 1 }
    }
    return out
}
